File: app/src/main/java/ru/ifmo/md/lesson8/database/DatabaseHelper.java
package ru.ifmo.md.lesson8.database;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.provider.BaseColumns;

/**
 * Created by Svet on 30.11.2014.
 */
public class DatabaseHelper  extends SQLiteOpenHelper implements BaseColumns {

    public final String TABLE_NAME = "WheatherData";

    //5 columns
    public final String CITY_NAME = "CityName";
    public final String TEMPERATURE_TODAY = "TemperatureToday";
    public final String TEMPERATURE_TOMORROW  = "TemperatureTomorrow";
    public final String TEMPERATURE_AFTER = "TemperatureAfter";

    public final String DESCRIPTION_TODAY = "DescriptionToday";
    public final String DESCRIPTION_TOMORROW = "DescriptionTomorrow";
    public final String DESCRIPTION_AFTER = "DescriptionAfter";

    public final String HUMIDITY = "Humidity";
    public final String PRESSURE = "Pressure";
    public final String WIND_SPEED = "WindSpeed";

    public final String SRC_TODAY = "SrcToday";
    public final String SRC_TOMORROW = "SrcTomorrow";
    public final String SRC_AFTER = "SrcAfter";

    public DatabaseHelper(Context context, String name, SQLiteDatabase.CursorFactory factory, int version) {
        super(context, name, factory, version);
    }

    @Override
    public void onCreate(SQLiteDatabase sqLiteDatabase) {
        String createTableScript = " create table " + TABLE_NAME + "( " +
                BaseColumns._ID + " integer primary key autoincrement, " +
                CITY_NAME + " text not null, " +

                TEMPERATURE_TODAY + " real, " +
                TEMPERATURE_TOMORROW + " real, " +
                TEMPERATURE_AFTER + " real, " +

                DESCRIPTION_TODAY + " text not null, " +
                DESCRIPTION_TOMORROW + " text not null, " +
                DESCRIPTION_AFTER + " text not null, " +

                HUMIDITY + " integer, " +
                PRESSURE + " integer, " +
                WIND_SPEED + " real, " +

                SRC_TODAY + " text not null, " +
                SRC_TOMORROW + " text not null, " +
                SRC_AFTER + " text not null);";
        sqLiteDatabase.execSQL(createTableScript);

    }

    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i2) {

    }
}

File: app/src/main/scala/me/volhovm/mweather/DatabaseHelper.scala
package me.volhovm.mweather

import java.util.Date

import android.content.{ContentResolver, ContentValues, Context}
import android.database.Cursor
import android.database.sqlite.{SQLiteDatabase, SQLiteOpenHelper}
import android.provider.BaseColumns

object DatabaseHelper extends BaseColumns {
  private val DATABASE_VERSION: Int = 1
  val DATABASE_NAME = "weather.db"
  val WEATHER_TABLE_NAME = "weather"
  val WEATHER_CITY = "weather_city"
  val WEATHER_COUNTRY = "weather_country"
  val WEATHER_HIGH_TEMP = "weather_high_temp"
  val WEATHER_LOW_TEMP = "weather_low_temp"
  val WEATHER_STATE_CODE = "weather_state_code"
  val WEATHER_STATE_DESC = "weather_state_descr"
  val WEATHER_HUMIDITY = "weather_humidity"
  val WEATHER_PRESSURE = "weather_pressure"
  val WEATHER_WIND = "weather_wind"
  val WEATHER_DATE = "weather_date"

  val CITIES_TABLE_NAME = "cities"
  val CITY = "city"

  val CREATE_WEATHER_TABLE = "create table " +
    WEATHER_TABLE_NAME + " (" +
    BaseColumns._ID + " integer primary key autoincrement, " +
    WEATHER_CITY + " text not null, " +
    WEATHER_COUNTRY + " text not null, " +
    WEATHER_HIGH_TEMP + " integer, " +
    WEATHER_LOW_TEMP + " integer, " +
    WEATHER_STATE_CODE + " integer, " +
    WEATHER_STATE_DESC + " text not null, " +
    WEATHER_HUMIDITY + " integer, " +
    WEATHER_PRESSURE + " integer, " +
    WEATHER_WIND + " text not null, " +
    WEATHER_DATE + " integer);"

  val CREATE_CITIES_TABLE = "create table " + CITIES_TABLE_NAME + " (" +
    BaseColumns._ID + " integer primary key autoincrement, " +
    CITY + " text not null);"
}

class DatabaseHelper(context: Context) extends SQLiteOpenHelper(context, null, null, 1) with BaseColumns {
  private val mContentResolver: ContentResolver = context.getContentResolver

  import me.volhovm.mweather.DatabaseHelper._

  override def onCreate(db: SQLiteDatabase): Unit = {
    db.execSQL(CREATE_CITIES_TABLE)
    db.execSQL(CREATE_WEATHER_TABLE)
  }

  override def onUpgrade(p1: SQLiteDatabase, p2: Int, p3: Int): Unit = throw new UnsupportedOperationException("CANNOT UPGRADE DB")

  def addWeather(weather: Weather): Unit = mContentResolver.insert(WeatherProvider.MAIN_CONTENT_URI, weather.getValues)

  def getCities(): List[String] = {
    var cursor: Cursor =
      mContentResolver.query(WeatherProvider.CITIES_CONTENT_URI, Array(DatabaseHelper.CITY), null, null, null)
    cursor.moveToFirst()
    compose(cursor, (curs: Cursor) => curs.getString(0))
  }

  def getWeatherByCity(cityname: String) = {
    val cursor: Cursor =
      mContentResolver.query(WeatherProvider.MAIN_CONTENT_URI, null, WEATHER_CITY + "='" + cityname + "'", null, null)
    cursor.moveToFirst()
    compose(cursor, cursorToWeather).reverse
  }

  def addCity(cityname: String) = {
    val values = new ContentValues()
    values.put(DatabaseHelper.CITY, cityname)
    mContentResolver.insert(WeatherProvider.CITIES_CONTENT_URI, values)
  }

  def deleteCity(cityname: String) = mContentResolver.delete(WeatherProvider.CITIES_CONTENT_URI, DatabaseHelper.CITY + "='" + cityname + "'", null)

  def updateCity(oldName: String, newName: String) = {
    val values = new ContentValues()
    values.put(DatabaseHelper.CITY, newName)
    mContentResolver.update(WeatherProvider.CITIES_CONTENT_URI, values, DatabaseHelper.CITY + "='" + oldName + "'", null)
  }

  private def compose[A](cursor: Cursor, foo: (Cursor) => A): List[A] =
    if (cursor.isAfterLast) Nil
    else foo(cursor) :: compose({
      cursor.moveToNext();
      cursor
    }, foo)

  def cursorToWeather(cursor: Cursor): Weather =
    if (cursor.isAfterLast)
      null
    else new Weather(
      cursor.getString(1),
      cursor.getString(2),
      (cursor.getInt(3), cursor.getInt(4)),
      new WeatherState(cursor.getInt(5), cursor.getString(6)),
      cursor.getInt(7).toDouble / 100,
      cursor.getInt(8),
      cursor.getString(9),
      new Date(1000l * cursor.getInt(10).toLong)
    )
}